internal/olm/operator/manager.go

```diff
@@ -0,0 +1,340 @@
+	if err = m.registryUp(ctx, olmresourceclient.OLMNamespace); err != nil {
+		return err
+	}
+	if !m.hasCatalogSource() {
+		registryGRPCAddr := opinternal.GetRegistryServiceAddr(pkgName, olmresourceclient.OLMNamespace)
+		catsrc := newCatalogSource(pkgName, m.namespace, withGRPC(registryGRPCAddr))
+		m.olmObjects = append(m.olmObjects, catsrc)
+	}
+	if !m.hasSubscription() {
+		channel, err := getChannelNameForCSVName(pkg, csv.GetName())
+		if err != nil {
+			return err
+		}
+		sub := newSubscription(pkgName, m.namespace, withChannel(channel),
+			withCatalogSource(getCatalogSourceName(pkgName), m.namespace))
+		m.olmObjects = append(m.olmObjects, sub)
+	}
+	if !m.hasOperatorGroup() {
+		// TODO(estroz): check if OG needs to be created in m.namespace first. If
+		// there is, CSV creation will fail with reason TooManyOperatorGroups. If
+		// the CSV's installModes don't support the target namespace selection of
+		// the OG, the CSV will fail with UnsupportedOperatorGroup. Might need to
+		// check these conditions before creating.
+		// https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/operatorgroups.md
+		olmCSV, err := registryutil.BundleCSVToCSV(csv)
+		if err != nil {
+			return err
+		}
+		targetNamespaces := []string{}
+		if csvSingleNamespace(olmCSV) || csvOwnNamespace(olmCSV) {
+			targetNamespaces = append(targetNamespaces, m.namespace)
+		}
+		og := newOperatorGroup(pkgName, m.namespace, targetNamespaces...)
+		m.olmObjects = append(m.olmObjects, og)
+	}
+	if err = m.client.DoCreate(ctx, m.olmObjects...); err != nil {
+		return err
+	}
+	nn := types.NamespacedName{
+		Name:      csv.GetName(),
+		Namespace: m.namespace,
+	}
+	log.Printf("Waiting for ClusterServiceVersion %q to reach 'Succeeded' phase", nn)
+	if err = m.client.DoCSVWait(ctx, nn); err != nil {
+		return err
+	}
+
+	status := m.status(ctx, bundle.Objects...)
+	if len(status.Resources) != len(bundle.Objects) {
+		return errors.Errorf("some operator %q resources did not install\n%s", csv.GetName(), status)
+	}
+	log.Infof("Successfully installed %q on OLM version %q", csv.GetName(), olmVer)
+	fmt.Print(status)
+
+	return nil
+}
+
+func (m *operatorManager) registryUp(ctx context.Context, namespace string) error {
+	rr := opinternal.RegistryResources{
+		Client:    m.client,
+		Manifests: m.manifests,
+	}
+	registryStale, err := rr.RegistryDataStale(ctx, namespace)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrap(err, "error checking registry data")
+		}
+		// ConfigMap doesn't exist yet, so create it as usual.
+		if err = rr.CreateRegistryManifests(ctx, namespace); err != nil {
+			return errors.Wrap(err, "error registering bundle")
+		}
+		return nil
+	}
+	if !registryStale && !m.force {
+		log.Printf("Registry data is current")
+		return nil
+	}
+	if m.force {
+		log.Printf("Forcefully recreating registry")
+	} else {
+		log.Printf("Registry data stale. Recreating registry")
```

Comment:
Right now, it seems like the only difference between `--force=true` and `--force=false` is that in the former case, we'll delete and re-create the registry even when the registry data is correct.

Is there a use case for that? Maybe to get a newer registry image version?

Also, without `--force`, we still delete and re-create when the data is stale. Is there a use case that dictates that we should bail out with stale data and require the `--force` flag?

Reply:
In this particular case `--force=true` isn't really useful, because as you say its not doing anything that a cache miss wouldn't.

I can't think of any case where we wouldn't want to rebuild the registry if its data is stale.

One slightly related thing I just thought of is that we don't necessarily have to re-create a `Subscription` and `CatalogSource` when rebuilding the registry, since they'll be pointing to the same registry server address. WDYT?